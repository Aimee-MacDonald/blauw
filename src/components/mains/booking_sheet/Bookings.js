import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {selectBooking} from '../../../state_management/actions/bookings'

const Bookings = props => (
  <StyledBookings onScroll={e => props.setScrollOffsets({x: e.target.scrollLeft, y: e.target.scrollTop})}>
    {props.bookings.map(({id, name, date, room, nights}) => (
      <StyledBooking
        key={id}
        date={date}
        room={room}
        nights={nights}
        onClick={() => props.dispatch(selectBooking(id))}
        selected={props.selectedBooking === id ? true : false}
      >{name}</StyledBooking>))}
  </StyledBookings>
)

const StyledBookings = styled.div`
  background-color: aquamarine;
  width: 100%;
  height: 100%;
  display: grid;
  grid-template-columns: repeat(60, 3.2rem);
  grid-template-rows:repeat(80, 2rem);
  overflow-x: scroll;
  overflow-y: scroll;
`

const StyledBooking = styled.div`
  background-color: ${props => props.selected ? 'orangered' : 'cornflowerblue'};
  grid-column-start: ${props => props.date};
  grid-column-end: ${props => props.date + props.nights};
  grid-row-start: ${props => props.room};
  display: flex;
  justify-content: center;
  align-items: center;
  border-radius: 0.2rem;
  margin: 0.1rem;
  transition: 0.4s;
  cursor: pointer;
`

const mapStateToProps = ({bookings}) => bookings
export default connect(mapStateToProps)(Bookings)