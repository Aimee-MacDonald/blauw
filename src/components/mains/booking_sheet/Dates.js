import React, {useEffect, useRef} from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {formatDate, normaliseDate} from '../../../util/util'

import {setCurrentDate, setStartDate, setEndDate, repopulateDates} from '../../../state_management/actions/dates'

const Dates = props => {
  const datesEl = useRef(null)

  useEffect(() => datesEl.current.scrollLeft = props.scroll)
  useEffect(() => {
    const dateObject = new Date()
    props.dispatch(setCurrentDate(normaliseDate(dateObject)))
    dateObject.setDate(dateObject.getDate() - 30)
    props.dispatch(setStartDate(normaliseDate(dateObject)))
    dateObject.setDate(dateObject.getDate() + 90)
    props.dispatch(setEndDate(normaliseDate(dateObject)))
    props.dispatch(repopulateDates())
  }, [])

  return(
    <StyledDates ref={datesEl}>
      {props.dates.dates.map((datestamp, index) => <StyledDate key={index} hoveredCell={props.hoveredCell === index + 1}>
        <p>{formatDate(datestamp).month}</p>
        <p>{formatDate(datestamp).date}</p>
      </StyledDate>)}
    </StyledDates>
  )
}

const StyledDates = styled.div`
  grid-column-start: 2;
  width: 100%;
  overflow-x: hidden;
  display: grid;
  grid-template-columns: repeat(91, 64px);
`

const StyledDate = styled.div`
  background-color: ${props => props.hoveredCell ? 'lime' : 'purple'};
  display: flex;
  justify-content: center;
  flex-flow: nowrap column;
  justify-content: center;

  p{
    text-align: center;
    font-weight: bold;
    font-size: 16pt;
    display: flex;
    align-items: center;
    justify-content: center;
  }
`

const mapStateToProps = ({dates}) => ({dates})

export default connect(mapStateToProps)(Dates)