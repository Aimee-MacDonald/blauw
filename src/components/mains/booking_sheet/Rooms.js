import React, {useRef, useEffect} from 'react'
import styled from 'styled-components'

const Rooms = props => {
  const roomsEl = useRef(null)
  useEffect(() => roomsEl.current.scrollTop = props.scroll)

  const rooms = [
    'Camping 1',
    'Camping 2',
    'Camping 3',
    'Camping 4',
    'Camping 5',
    'Camping 6',
    'Camping 7',
    'Camping 8',
    'Camping 9',
    'Camping 10',
    'Camping 11',
    'Camping 12',
    'Camping 13',
    'Camping 14',
    'Camping 15',
    'Camping 16',
    'Camping 17',
    'Camping 18',
    'Camping 19',
    'Camping 20',
    'Camping 21',
    'Camping 22',
    'Camping 23',
    'Camping 24',
    'Camping 25',
    'Camping 26',
    'Camping 27',
    'Camping 28',
    'Camping 29',
    'Camping 30',
    'Camping 31',
    'Camping 32',
    'Camping 33',
    'Camping 34',
    'Camping 35',
    'Camping 36',
    'Camping 37',
    'Camping 38',
    'Camping 39',
    'Camping 40',
    'Camping 41',
    'Camping 42',
    'Camping 43',
    'Camping 44',
    'Camping 45',
    'Camping 46',
    'Camping 47',
    'Camping 48',
    'Camping 49',
    'Camping 50',
    'Camping 51',
    'Camping 52',
    'Camping 53',
    'Camping 54',
    'Camping 55',
    'Camping 56',
    'Camping 57',
    'Camping 58',
    'Camping 59',
    'Camping 60',
    'Camping 61',
    'Camping 62',
    'Camping 63',
    'Camping 64',
    'Camping 65',
    'Camping 66',
    'Camping 67',
    'Camping 68',
    'Camping 69',
    'Camping 70',
    'Camping 71',
    'Camping 72',
    'Camping 73',
    'Camping 74',
    'Camping 75',
    'Camping 76',
    'Camping 77',
    'Camping 78',
    'Camping 79',
    'Camping 80'
  ]
  
  return(
    <StyledRooms ref={roomsEl}>
      {rooms.map((room, i) => <StyledRoom key={i} hoveredCell={props.hoveredCell === i + 1}><p>{room}</p></StyledRoom>)}
    </StyledRooms>
  )
}

const StyledRooms = styled.div`
  height: 100%;
  overflow-y: hidden;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: repeat(80, 32px);
`

const StyledRoom = styled.div`
  background-color: ${props => props.hoveredCell ? 'lime' : 'purple'};
  display: flex;
  align-items: center;

  p{
    padding-left: 0.4rem;
  }
`

export default Rooms