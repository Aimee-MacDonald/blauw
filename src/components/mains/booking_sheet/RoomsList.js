import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getRoomGroupings} from '../../../state_management/selectors/rooms'

export const RoomsList = ({groups, groupFlags, setGroupFlags}) => (
  <StyledRoomsList>
    {groups.map(({groupName, rooms}) => (
      <StyledGroup key={groupName}>
        <h3 onClick={e => setGroupFlags({
          ...groupFlags,
          [groupName]: {
            ...groupFlags[groupName],
            open: !groupFlags[groupName].open
          }
        })}>{groupName}</h3>

        {groupFlags[groupName].open && rooms.map(({roomName, maxPax}) => {
          const beds = []
          for(let i = 0; i < maxPax; i++) beds.push(`Bed ${i + 1}`)

          return(
            <StyledRooms key={roomName}>
              <h3 onClick={e => setGroupFlags({
                ...groupFlags,
                [groupName]: {        
                  ...groupFlags[groupName],
                  rooms: {
                    ...groupFlags[groupName].rooms,
                    [roomName]: !groupFlags[groupName].rooms[roomName]
                  }
                }
              })}>{roomName}</h3>

              {groupFlags[groupName].rooms[roomName] && beds.map(bed => (
                <StyledBed key={`${roomName}${bed}`}>
                  <h3>{bed}</h3>
                </StyledBed>
              ))}
            </StyledRooms>
          )
        })}
      </StyledGroup>
    ))}
  </StyledRoomsList>
)

const StyledRoomsList = styled.div`
  background-color: lime;
  height: 100%;
  overflow-y: hidden;
`

const StyledGroup = styled.div`
  background-color: orangered;

  h3{
    height: 32px;
    display: grid;
    place-items: center center;
    cursor: pointer;
  }
`

const StyledRooms = styled.div`
  background-color: magenta;
`

const StyledBed = styled.div`
  background-color: violet;

  h3{
    place-items: center right;
    padding-right: 1rem;
  }
`

const mapStateToProps = ({rooms}) => ({groups: getRoomGroupings(rooms)})

export default connect(mapStateToProps)(RoomsList)