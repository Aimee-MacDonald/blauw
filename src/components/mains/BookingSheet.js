import React from 'react'
import styled from 'styled-components'

import Dates from './booking_sheet/Dates'

const BookingSheet = () => (
  <StyledBookingSheet>
    <Dates/>
  </StyledBookingSheet>
)

const StyledBookingSheet = styled.div`
  background-color: indigo;
  width: 100%;
  height: 100%;
`

export default BookingSheet