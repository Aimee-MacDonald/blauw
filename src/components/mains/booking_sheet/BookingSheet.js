import React, {useState} from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getRoomGroupings} from '../../../state_management/selectors/rooms'

import Bookings from './Bookings'
import Dates from './Dates'
import RoomsList from './RoomsList'

export const BookingSheet = props => {
  let defaultRoomsFlags = []

  props.groups.forEach(({groupName, rooms}) => {
    let roomFlags = {}
    rooms.forEach(room => roomFlags = {
      ...roomFlags,
      [room.roomName]: false
    })

    defaultRoomsFlags = {
      ...defaultRoomsFlags,
      [groupName]: {
        open: false,
        rooms: roomFlags
      }
    }
  })

  const [scrollOffsets, setScrollOffsets] = useState({x: 0, y: 0})
  const [mouseCoordinates, setMouseCoordinates] = useState({x: null, y: null})
  const [hoveredCell, setHoveredCell] = useState({x: null, y: null})
  const [roomsFlags, setRoomsFlags] = useState(defaultRoomsFlags)

  const handleMouseMove = e => {
    setMouseCoordinates({x: e.clientX, y: e.clientY})
    if(mouseCoordinates.x < 256 || mouseCoordinates.y < 64){
      setHoveredCell({x: null, y: null})
    } else {
      setHoveredCell({
        x: parseInt((mouseCoordinates.x - 256 + scrollOffsets.x) / (window.innerWidth - 256) * ((window.innerWidth - 256) / 64)) + 1,
        y: parseInt((mouseCoordinates.y - 64 + scrollOffsets.y) / (window.innerHeight - 64) * ((window.innerHeight - 64) / 32)) + 1
      })
    }
  }

  return(
    <StyledBookingSheet onMouseMove={e => handleMouseMove(e)}>
      <Dates scroll={scrollOffsets.x} hoveredCell={hoveredCell.x}/>
      <RoomsList scroll={scrollOffsets.y} hoveredCell={hoveredCell.y} groupFlags={roomsFlags} setGroupFlags={setRoomsFlags}/>
      <Bookings setScrollOffsets={setScrollOffsets}/>
    </StyledBookingSheet>
  )
}

const StyledBookingSheet = styled.div`
  background-color: indigo;
  width: 100%;
  height: 100%;
  display: grid;
  grid-template-columns: 256px 1fr;
  grid-template-rows: 64px 1fr;
`

const mapStateToProps = ({rooms}) => ({groups: getRoomGroupings(rooms)})

export default connect(mapStateToProps)(BookingSheet)