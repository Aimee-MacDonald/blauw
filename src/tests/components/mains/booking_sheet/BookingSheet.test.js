import React from 'react'
import {shallow} from 'enzyme'

import {BookingSheet} from '../../../../components/mains/booking_sheet/BookingSheet'

test('Render default booking sheet', () => {
  const wrapper = shallow(<BookingSheet/>)
  expect(wrapper.getElement()).toMatchSnapshot()
})