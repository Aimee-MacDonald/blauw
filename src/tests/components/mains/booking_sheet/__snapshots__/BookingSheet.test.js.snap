// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render default booking sheet 1`] = `
<styled.div
  onMouseMove={[Function]}
>
  <Memo(Connect(Dates))
    hoveredCell={null}
    scroll={0}
  />
  <Memo(Connect(RoomsList))
    groupFlags={
      Object {
        "Camping": Object {
          "open": false,
          "rooms": Object {
            "camping": false,
          },
        },
        "Dorms": Object {
          "open": false,
          "rooms": Object {
            "dove_dorm": false,
            "forest_house": false,
            "forest_house_2": false,
            "loerie_dorm": false,
            "owl_dorm": false,
          },
        },
        "Dove": Object {
          "open": false,
          "rooms": Object {
            "dove_left": false,
            "dove_middle": false,
            "dove_right": false,
          },
        },
        "Loerie": Object {
          "open": false,
          "rooms": Object {
            "loerie_family": false,
            "loerie_left": false,
            "loerie_right": false,
          },
        },
        "Owl": Object {
          "open": false,
          "rooms": Object {
            "owl_left": false,
            "owl_lounge": false,
            "owl_right": false,
          },
        },
        "Safari Tents": Object {
          "open": false,
          "rooms": Object {
            "safari_tent_1": false,
            "safari_tent_10": false,
            "safari_tent_11": false,
            "safari_tent_12": false,
            "safari_tent_2": false,
            "safari_tent_3": false,
            "safari_tent_4": false,
            "safari_tent_5": false,
            "safari_tent_6": false,
            "safari_tent_7": false,
            "safari_tent_8": false,
            "safari_tent_9": false,
          },
        },
      }
    }
    hoveredCell={null}
    scroll={0}
    setGroupFlags={[Function]}
  />
  <Memo(Connect(Bookings))
    setScrollOffsets={[Function]}
  />
</styled.div>
`;
